### ISSUE TITLE
SPades assembly issue

### ISSUE THREAD

**Issue by hutumpyancha** [faircloth-lab/phyluce#250]
Hello, 

I am currently working with UCE data (from octocoral samples) using the Phyluce pipeline on LONI and trying to assemble my reads using spades. Out of my total 52 samples, 11 are failing to assemble. I have tried running them seperately and am still unable to make them work. Initially when I was running them the log file would have the following line -
[spades_error.log](https://github.com/faircloth-lab/phyluce/files/7253771/spades_error.log)

0:15:59.443 324M / 964M ERROR K-mer Counting (kmer_data.cpp : 349) The reads contain too many k-mers to fit into available memory. You need approx. 16.3775GB of free RAM to assemble your dataset
[spades-1.log](https://github.com/faircloth-lab/phyluce/files/7253772/spades-1.log)

Later when I increased the number of cores I was using, it still failed to work.  I am attaching the script, configuration file, error and log file(of one sample). This time, the log file has the follwing lines -

-[38;21m2021-09-22 16:24:46,757 - phyluce_assembly_assemblo_spades - INFO - Argument --memory: 8
-mimalloc: error: unable to allocate OS memory (268435456 bytes, error code: 12 [Cannot allocate memory], address: (nil), large only: 0, allow large: 1

I  am sure how to allocate more RAM to my SPades assembly job; so, I would really appreciate it if you could assist me with this issue. My other samples seem to have worked properly and I cannot figure out what alterations need to be made to make these sample assemble, 

Regards, Upasana

The latest script I used was as follows - 

#!/bin/bash
#PBS -l nodes=1:ppn=20
#PBS -l walltime=72:00:00
#PBS -N spades
#PBS -o spades.log
#PBS -e spades_error.log
#PBS -q checkpt
#PBS -A loni_deepsea2
#PBS -m abe
#PBS -M upasana.ganguly1@louisiana.edu
#
eval "$(conda shell.bash hook)"
conda activate phyluce-1.7.1
#
#
#
cd /work/upasanag/spades
phyluce_assembly_assemblo_spades --conf /work/upasanag/spades/assembly.conf  \ 
--output /work/upasanag/spades/spades-assemblies \
--cores 20

The log files - 

[spades_error.log](https://github.com/faircloth-lab/phyluce/files/7253778/spades_error.log)

[spades-1.log](https://github.com/faircloth-lab/phyluce/files/7253781/spades-1.log)

**Comment by brantfaircloth** [comment]
You should be able to adjust this using `--memory`.  Assuming you have 64 GB available, that would look like:
```
#!/bin/bash
#PBS -l nodes=1:ppn=20
#PBS -l walltime=72:00:00
#PBS -N spades
#PBS -o spades.log
#PBS -e spades_error.log
#PBS -q checkpt
#PBS -A loni_deepsea2
#PBS -m abe
#PBS -M upasana.ganguly1@louisiana.edu

eval "$(conda shell.bash hook)"
conda activate phyluce-1.7.1

cd /work/upasanag/spades
phyluce_assembly_assemblo_spades --conf /work/upasanag/spades/assembly.conf \
--output /work/upasanag/spades/spades-assemblies \
--cores 20 \
--memory 64
```

**Comment by hutumpyancha** [comment]
Thank you so much!! I will try it this way.

**Comment by brantfaircloth** [comment]
Depending on how much read data you have, you might need to move to QBC.

**Comment by hutumpyancha** [comment]
Of the 11 samples that did not work, the sample with the highest number of reads has 19024218 reads and the lowest number is 1528240.

**Comment by hutumpyancha** [comment]
I tried with the altered script (added --memory 64)on one sample but it failed to work again. I am including the phyluce_assembly_assemblo_spades.log and the spades.log files here; line 12 of the former states that it cannot find the expected assembly files and line 72 of the latter again states error code 12: cannot allocate memory
[phyluce_assembly_assemblo_spades.log](https://github.com/faircloth-lab/phyluce/files/7255386/phyluce_assembly_assemblo_spades.log)
.
[spades.log](https://github.com/faircloth-lab/phyluce/files/7255388/spades.log)

**Comment by brantfaircloth** [comment]
If it’s working for some samples and not others, it’s most likely a RAM issue.  you could try to downsample your reads for the problematic samples:

http://protocols.faircloth-lab.org/en/latest/protocols-computer/snippets/random-computer-snippets.html#subsample-reads-for-r1-and-r2-using-seqtk

**Comment by brantfaircloth** [comment]
Also, from the logs, it looks like you used `—memory 8`

**Comment by hutumpyancha** [comment]
#!/bin/bash
#PBS -l nodes=1:ppn=20
#PBS -l walltime=72:00:00
#PBS -N spades
#PBS -o spades.log
#PBS -e spades_error.log
#PBS -q checkpt
#PBS -A loni_deepsea2
#PBS -m abe
#PBS -M upasana.ganguly1@louisiana.edu
#
eval "$(conda shell.bash hook)"
conda activate phyluce-1.7.1
#
#
#
cd /work/upasanag/spades
phyluce_assembly_assemblo_spades --conf /work/upasanag/spades/assembly.conf  \ 
--output /work/upasanag/spades/spades-assemblies \
--cores 20 \
--memory 64

This is the script that I used. Even though it says in the script --memory 64, I too noticed that it says 8 in the log. I cannor understand why that is happening.

**Comment by brantfaircloth** [comment]
The formatting on your submission script is making it hard to read. Are there slashes after “spades-assemblies” and “20”?  There should be.  Also, on which LONI machine is this?

**Comment by hutumpyancha** [comment]
Yes, there are slashes after both "spades-assemblies" and "20"; the formatting of the script text got changed when I was copying it in the message, apologies for that. This is on qb3.

**Comment by brantfaircloth** [comment]
I'm not sure why this is not working - it seems to work fine for me when testing (and has worked fine for other folks who have tried the same).  You might try running the script interactively to see if something about the phyluce command in the file you are sending to `qsub` is causing the issue.

Alternatively, you can run spades *outside* of phyluce - you'll just need to do this for each set of samples, then symlink them into the "contigs" directory.  The following is equivalent to what phyluce runs:
```
spades.py --careful --sc --memory 60 --threads 20 --cov-cutoff 5 --pe1-1 <path to r1 files> --pe1-2 <path to r2 files> --pe1-s <path to singleton files> -o <output dir name>
```

**Comment by hutumpyancha** [comment]
okay..I will try running spades in these two ways and see if I can make these samples successfully assemble; thank you so much!
